[WGSL] Gardening of WGSL parser unit tests

djg opened the pull request:
#### 0a4bde8f1d82ae6a0a5fa0db6ef1855c485f45e5
<pre>
[WGSL] Gardening of WGSL parser unit tests
<a href="https://bugs.webkit.org/show_bug.cgi?id=245654">https://bugs.webkit.org/show_bug.cgi?id=245654</a>
rdar://problem/100397595

Reviewed by Myles C. Maxfield.

Gardening of unit tests, including:
- Add full parsing tests for complete sample vertex and fragment shaders.
- Use XCTAssertTrue &amp; XCTAssertEqual instead of XCTAssert.
- Refactor tests to use auto&amp; instead of doubling up on type names.
- Add functions for common AST checks.

* Source/WebGPU/WGSLUnitTests/WGSLParserTests.mm:
(-[WGSLParserTests testParsingStruct]):
(-[WGSLParserTests testParsingGlobalVariable]):
(-[WGSLParserTests testParsingFunctionDecl]):
(-[WGSLParserTests testTrivialGraphicsShader]):
(-[WGSLParserTests testParsingLocalVariable]):
(-[WGSLParserTests testParsingArrayAccess]):
(-[WGSLParserTests testParsingUnaryExpression]):
(-[WGSLParserTests testParsingTriangleVert]):
(-[WGSLParserTests testParsingRedFrag]):

Canonical link: <a href="https://commits.webkit.org/254932@main">https://commits.webkit.org/254932@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d41802c8de6d499346e1e69898ed9ec5e6ad536a

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/90629 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/69/builds/35211 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/21241 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/99948 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/158289 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/94639 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/64/builds/33708 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/28868 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/83000 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/96367 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/96284 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/68/builds/26829 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/77460 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/26674 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/81607 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/81389 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/69702 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/34804 "Built successfully") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/15426 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/32615 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/16405 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/3441 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/36381 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/39334 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/38302 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/35493 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

litherum commented on Source/WebGPU/WGSLUnitTests/WGSLParserTests.mm:
```diff
@@ -42,24 +42,24 @@
 #define ASSERT_INT_LITERAL(node, int_value) \
     do { \
         XCTAssertTrue((node).isAbstractIntLiteral()); \
-        WGSL::AST::AbstractIntLiteral& intLiteral = downcast<WGSL::AST::AbstractIntLiteral>(node); \
```
Why do we have an AST namespace?

djg replied:
I don't know.

litherum commented on Source/WebGPU/WGSLUnitTests/WGSLParserTests.mm:
```diff
@@ -175,64 +175,64 @@ - (void)testTrivialGraphicsShader {
-        WGSL::AST::ReturnStatement& stmt = downcast<WGSL::AST::ReturnStatement>(func.body().statements()[0].get());
-        XCTAssert(stmt.maybeExpression());
-        XCTAssert(stmt.maybeExpression()->isIdentifier());
+        auto& func = shader->functions()[0].get();
+        XCTAssertEqual(func.attributes().size(), 1u);
+        XCTAssertTrue(func.attributes()[0]->isStage());
+        XCTAssertEqual(downcast<WGSL::AST::StageAttribute>(func.attributes()[0].get()).stage(), WGSL::AST::StageAttribute::Stage::Vertex);
+        XCTAssertEqual(func.name(), "vertexShader"_s);
+        XCTAssertEqual(func.parameters().size(), 1u);
+        XCTAssertEqual(func.parameters()[0]->name(), "x"_s);
+        XCTAssertEqual(func.parameters()[0]->attributes().size(), 1u);
+        XCTAssertTrue(func.parameters()[0]->attributes()[0]->isLocation());
+        XCTAssertEqual(downcast<WGSL::AST::LocationAttribute>(func.parameters()[0]->attributes()[0].get()).location(), 0u);
+        XCTAssertTrue(func.parameters()[0]->type().isParameterized());
+        auto& paramType = downcast<WGSL::AST::ParameterizedType>(func.parameters()[0]->type());
+        XCTAssertEqual(paramType.base(), WGSL::AST::ParameterizedType::Base::Vec4);
+        XCTAssertTrue(paramType.elementType().isNamed());
+        XCTAssertEqual(downcast<WGSL::AST::NamedType>(paramType.elementType()).name(), "f32"_s);
+        XCTAssertEqual(func.returnAttributes().size(), 1u);
+        XCTAssertTrue(func.returnAttributes()[0]->isBuiltin());
+        XCTAssertEqual(downcast<WGSL::AST::BuiltinAttribute>(func.returnAttributes()[0].get()).name(), "position"_s);
+        XCTAssertTrue(func.maybeReturnType());
+        XCTAssertTrue(func.maybeReturnType()->isParameterized());
+        XCTAssertEqual(func.body().statements().size(), 1u);
+        XCTAssertTrue(func.body().statements()[0]->isReturn());
+        auto& stmt = downcast<WGSL::AST::ReturnStatement>(func.body().statements()[0].get());
+        XCTAssertTrue(stmt.maybeExpression());
+        XCTAssertTrue(stmt.maybeExpression()->isIdentifier());
```
(I'm not looking at these things very hard, I assume they match)

litherum commented on Tools/TestWebKitAPI/Tests/WGSL/ParserTests.cpp:
```diff
@@ -39,6 +39,32 @@
 #import "WGSL.h"
 #import <wtf/DataLog.h>
 
+#define ASSERT_BUILTIN(attr, attr_name) \
```
why macro? why not function?

you can use `__file__` and `__line__` if necessary

litherum commented on Tools/TestWebKitAPI/Tests/WGSL/ParserTests.cpp:
```diff
@@ -341,15 +366,111 @@ TEST(WGSLParserTests, UnaryExpression)
+        "}\n"_s);
+
+    if (!shader.has_value())
+        dataLogLn(shader.error());
```
Can we use `WTFLogAlways()` throughout WebGPU/WGSL, instead of `dataLogLn`?

litherum commented on Tools/TestWebKitAPI/Tests/WGSL/ParserTests.cpp:
```diff
@@ -341,15 +366,111 @@ TEST(WGSLParserTests, UnaryExpression)
+    if (!shader.has_value())
+        dataLogLn(shader.error());
+    EXPECT_TRUE(shader.has_value());
+    EXPECT_TRUE(shader->directives().isEmpty());
+    EXPECT_TRUE(shader->structs().isEmpty());
+    EXPECT_TRUE(shader->globalVars().isEmpty());
+    EXPECT_EQ(shader->functions().size(), 1u);
```
🆒

djg committed:
[WGSL] Gardening of WGSL parser unit tests
https://bugs.webkit.org/show_bug.cgi?id=245654
rdar://problem/100397595

Reviewed by Myles C. Maxfield.

Gardening of unit tests, including:
- Add full parsing tests for complete sample vertex and fragment shaders.
- Use XCTAssertTrue & XCTAssertEqual instead of XCTAssert.
- Refactor tests to use auto& instead of doubling up on type names.
- Add functions for common AST checks.

* Source/WebGPU/WGSLUnitTests/WGSLParserTests.mm:
(-[WGSLParserTests testParsingStruct]):
(-[WGSLParserTests testParsingGlobalVariable]):
(-[WGSLParserTests testParsingFunctionDecl]):
(-[WGSLParserTests testTrivialGraphicsShader]):
(-[WGSLParserTests testParsingLocalVariable]):
(-[WGSLParserTests testParsingArrayAccess]):
(-[WGSLParserTests testParsingUnaryExpression]):
(-[WGSLParserTests testParsingTriangleVert]):
(-[WGSLParserTests testParsingRedFrag]):

Canonical link: https://commits.webkit.org/254932@main
